### PR TITLE
Resurrect hairpin NAT

### PR DESCRIPTION
**Pull request by icecrime**
We're throwing a new attempt at removing the userland proxy from Docker. The blocking issue we discovered last time was that hairpin NAT was not available on some systems (see docker/docker#9134).

This PR reintroduces @phemmer work from #62 and adds the ability to enable Hairpin NAT on a per-network basis using the `Network.HairpinNAT` flag (disabled by default).

### CODE REVIEW

**Comment by mrunalp** [comment]
LGTM

**Comment by dqminh** [comment]
LGTM

File: network/types.go [hunk 1]
```diff
@@ -38,6 +38,12 @@ type Network struct {
 	// container's interfaces if a pair is created, specifically in the case of type veth
 	// Note: This does not apply to loopback interfaces.
 	TxQueueLen int `json:"txqueuelen,omitempty"`
+
+	// HairpinNAT specifies if hairpin NAT should be enabled on the virtual interface
+	// bridge port in the case of type veth
+	// Note: This is unsupported on some systems.
+	// Note: This does not apply to loopback interfaces.
+	HairpinNAT bool `json:"hairpin_nat"`
```
Comment (LK4D4) [inline network/types.go]: Actually that thing not called `HairpinNAT`. It is rather `HairpinMode` or just `Hairpin`

**Comment by icecrime** [comment]
Renamed `bool` `HairpinNAT` to `Hairpin`.

**Comment by icecrime** [comment]
Renamed `bool` `Hairpin` to `HairpinMode`. :tada:

**Comment by mrunalp** [comment]
Justification for name -- http://lwn.net/Articles/347344/

**Pull request merged**